src/librustc_mir/borrow_check/mod.rs

```diff
@@ -847,14 +881,57 @@ impl<'cx, 'gcx, 'tcx> MirBorrowckCtxt<'cx, 'gcx, 'tcx> {
         drop_place: &Place<'tcx>,
         erased_drop_place_ty: ty::Ty<'gcx>,
         span: Span,
+        prev_seen: SeenTy<'_, 'gcx>,
     ) {
+        if prev_seen.have_seen(erased_drop_place_ty) {
+            // if we have directly seen the input ty `T`, then we must
+            // have had some *direct* ownership loop between `T` and
+            // some directly-owned (as in, actually traversed by
+            // recursive calls below) part that is also of type `T`.
+            //
+            // Note: in *all* such cases, the data in question cannot
+            // be constructed (nor destructed) in finite time/space.
+            //
+            // Proper examples, some of which are statically rejected:
+            //
+            // * `struct A { field: A, ... }`:
+            //   statically rejected as infinite size
+            //
+            // * `type B = (B, ...);`:
+            //   statically rejected as cyclic
+            //
+            // * `struct C { field: Box<C>, ... }`
+            // * `struct D { field: Box<(D, D)>, ... }`:
+            //   *accepted*, though impossible to construct
+            //
+            // Here is *NOT* an example:
+            // * `struct Z { field: Option<Box<Z>>, ... }`:
+            //   Here, the type is both representable in finite space (due to the boxed indirection)
+            //   and constructable in finite time (since the recursion can bottom out with `None`).
+            //   This is an obvious instance of something the compiler must accept.
+            //
+            // Since some of the above impossible cases like `C` and
+            // `D` are accepted by the compiler, we must take care not
+            // to infinite-loop while processing them. But since such
+            // cases cannot actually arise, it is sound for us to just
+            // skip them during drop. If the developer uses unsafe
+            // code to construct them, they should not be surprised by
+            // weird drop behavior in their resulting code.
```

Comment:
Uh, can we not fall-back-to-error instead of fall-back-to-accept? Seems like a disaster waiting to happen.^^

Reply:
We cannot fallback to error, at least not to hard error. That could break hypothetical code.

We could issue a future compatibility warning.

Or we could *try* to emit drop code that error dynamically ... maybe that is what you are asking for?

(When I wrote "skip them during drop", I meant during this static analysis. I believe the actual dynamic behavior here will be to ... infinite loop ... i think... )

Reply:
([This is the test case](https://github.com/rust-lang/rust/blob/c02c00b8455d6ec6eff50ae94bebb4a424c95e02/src/test/ui/issue-45696-no-variant-box-recur.rs) showing the kind of case that concerns me. It is currently marked as `run-pass`; we could make a variant of it that actually conjures up one of this infinitely regressed things, but I'm pretty sure that the unsafe code guidelines group would tag such a test as having undefined behavior, no?)

Reply:
> I'm pretty sure that the unsafe code guidelines group would tag such a test as having undefined behavior,

Yeah, I agree.

> When I wrote "skip them during drop", I meant during this static analysis. I believe the actual dynamic behavior here will be to ... infinite loop ... i think... 

Probably it'll fill up the stack. ;)
What does it mean for the static analysis to skip this? Does it not consider this a use, or so? That seems problematic.^^

Reply:
It considers the first one it sees a use, but not the second one encountered via the recursive descent.

But to even get to that code during dynamic execution you’d have to already built the invalid thing and be witnessing UB.

Maybe I should make a more complete illustrative test case for us to discuss, with the aforementioned construction...?

Reply:
> It considers the first one it sees a use, but not the second one encountered via the recursive descent.

I see. Yes, without knowing any of the details, this does make some sense.